keras_cv/training/contrastive/contrastive_trainer.py

```diff
@@ -0,0 +1,260 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+
+class ContrastiveTrainer(keras.Model):
+    """Creates a self-supervised contrastive trainer for a model.
+
+    Args:
+        encoder: a `keras.Model` to be pre-trained. In most cases, this encoder
+            should not include a top dense layer.
+        augmenter: a preprocessing layer to randomly augment input images for contrastive learning,
+            or a tuple of two separate augmenters for the two sides of the contrastive pipeline.
+        projector: a projection model for contrastive training, or a tuple of two separate
+            projectors for the two sides of the contrastive pipeline.
+        include_probe: Whether to include a single fully-connected layer during
+            training for probing classification accuracy using the learned encoding.
+            Note that this should be specified iff training with labeled images.
+            If provided, `classes` must be provided.
+        probe_metrics: a list of metrics for the linear probe, only used if `include_probe` is True.
+        classes: optional number of classes to classify images into, only to be
+            specified if `include_probe` is True.
+
+    Returns:
+      A `keras.Model` instance.
+
+
+    Usage:
+    ```python
+    encoder = keras_cv.models.DenseNet121(include_rescaling=False, include_top=False, pooling="avg")
+    augmenter = keras_cv.layers.preprocessing.RandomFlip()
+    projector = keras.layers.Dense(64)
+
+    trainer = keras_cv.training.ContrastiveTrainer(
+        encoder,
+        augmenter,
+        projector
+    )
+
+    trainer.compile(
+        optimizer=keras.optimizers.Adam(),
+        loss=keras_cv.losses.SimCLRLoss()
+    )
+
+    unlabeled_images = load_data()
+    trainer.fit(unlabeled_images)
+    ```
+
+    """
+
+    def __init__(
+        self,
+        encoder,
+        augmenter,
+        projector,
+        include_probe,
+        classes=None,
+    ):
+        super().__init__()
+
+        if encoder.output.shape.rank != 2:
+            raise ValueError("Encoder must have a flattened output")
```

Comment:
try to provide actionable error messages with some information regarding what values the user passed:

something like:
```
raise ValueError(f"`encoder` must have a flattened output.  Expected rank(encoder.output.shape)=2, got rank(encoder.output.shape={encoder.output.shape}")
```

The goal of our error messages in KerasCV is to make it so users never go to the docs: the code fixes itself.

Reply:
Ack -- done.